### ISSUE TITLE
Problem: AM17 externals repos not available yet

### ISSUE THREAD

**Issue by sevein** [artefactual-labs/am-packbuild#80]
In AM16 we had the following repos:
- https://packages.archivematica.org/1.6.x/centos-extras 
- https://packages.archivematica.org/1.6.x/ubuntu-externals

We haven't created the AM17 externals repos yet. This should be done at some point so we know that the dependencies are installed and the right versions are used, like siegfried v1.7.6. Maybe that's the only change needed?

---

Related question that I'm trying to squeeze here hoping that you can respond. I don't really understand the mechanics of deb/rpm repositories. What's in the externals repo that needs to be namespaced as `/1.6.x` or `/1.7.x`? Couldn't a single repository contain multiple versions of a package? E.g. I see how multiple versions of the siegfried package coexist under the same https://packages.archivematica.org/1.6.x/centos-extras/ repository: `v1.5.0`, `v1.6.7` and `v1.7.3`? The archivematica-mcp-client package [depends on](https://github.com/artefactual-labs/am-packbuild/blob/8b80b030bcaf0861ca28b20f338e2acac06fdae7/rpm/archivematica/archivematica.spec#L58) `siegfried` but we don't define a version number constraint. If we started doing this, could we move toward a situation where we'd only need one `externals` repository that doesn't need to be versioned? Thank you in advance.

**Comment by scollazo** [comment]
Debian repositories made with reprepro will only publish the latest version available. I'm unsure about centos behaviour, but for sure, both systems will install the latest available version for a package.

The reasoning to keep different externals repositories for different versions, is to avoid forcing a package upgrade when new packages are added for new releases. If we add them to the same repository, environments using archivematica 1.6 will get packages targeted for am 1.7.

**Comment by sevein** [comment]
> If we add them to the same repository, environments using archivematica 1.6 will get packages targeted for am 1.7.

Wouldn't be possible to add version number constraints like explained [here](http://rpm.org/user_doc/more_dependencies.html). E.g. in my question I mentioned the following example:

```rpm
Requires: siegfried
```

Can't that be updated as follows?

```rpm
Requires: siegfried = 1.7.6
```

Would it be significantly easier if we had a single externals repo?

**Comment by scollazo** [comment]
No, we would have to exclude the package from being updated through yum as explained [here](https://access.redhat.com/solutions/10185) , or else new versions will get installed.

Also, if we agree that we need a different repo for each archivematica release, we might use the same repo for archivematica and external packages. Does this make sense?

**Comment by sevein** [comment]
> No, we would have to exclude the package from being updated through yum as explained here , or else new versions will get installed.

Where can I read more about this behavior that you're describing? Shouldn't the package manager ask the user to deal with the conflict instead of violating the constraint? That's very surprising.

> Also, if we agree that we need a different repo for each archivematica release, we might use the same repo for archivematica and external packages. Does this make sense?

Do you mean having a repo like `/1.7.0`? I actually like the `/1.7.x` approach better because we can treat it as a delivery channel for any patch release we want to release later under the AM17 umbrella, e.g. a possible `1.7.1` bugfix/patch if we wanted to.

If you think that having putting AM packages and externals in the same repo is going to simplify things then I'm all in. But what exactly we would be gaining with that approach? Would that enable us to pin deps or is just a matter of deduplication?

How would the package manager deal with the following situation: our archivematica-mcp-client 1.7.0 package depends on siegfried 1.7.6 provided by us within the same repo or an externals repo. The user adds later another repo that includes a newer version of siegfried - would it be eventually installed by the package manager during a system update?

I'm putting stress on the version number of the dependencies because it seems that's a concern in the digital preservation domain because different versions can produce different outputs. If this is going to be very hard to achieve I would also say it's ok to bump it to AM18.

**Comment by scollazo** [comment]
> Where can I read more about this behavior that you're describing? Shouldn't the package manager ask the user to deal with the conflict instead of violating the cfoonstraint? That's very surprising.

It's the standard behavior on almost any linux systems that uses package repositories. When a package with a higer version number version is available, it's installed as an update.

In debian systems, you can avoid this by setting a package "on hold" . You can check how to do it [here](http://www.debianadmin.com/how-to-prevent-a-package-from-being-updated-in-debian.html), and in my last comment I explain how that's done in redhat by using yum excludes.

Also, you brought pinning to the discussion, but I don't see it fit.  APT [pinning](https://wiki.debian.org/AptPreferences) works at a repository leveln, and is used to pull one or more packages from a later version, in order to access more up-to-date versions of packages. However, packages have dependencies which can conflict, so pinning (i.e., not pulling all your packages from one version) can cause problems. 

> How would the package manager deal with the following situation: our archivematica-mcp-client 1.7.0 package depends on siegfried 1.7.6 provided by us within the same repo or an externals repo. The user adds later another repo that includes a newer version of siegfried - would it be eventually installed by the package manager during a system update?

The answer is yes. If a package with a newer version is available, the upgrade will install it, as that's how security and any other updates gets added to the system. 

That's true for siegfried, but also for elasticsearch. ffmpeg, or any other package. If someone adds a repository with carries a higher version, it'll get installed unless pinning is set in place.

> Do you mean having a repo like /1.7.0? I actually like the /1.7.x approach better because we can treat its as a delivery channel for any patch release we want to release later under the AM17 umbrella, e.g. a possible 1.7.1 bugfix/patch if we wanted to.

I agree that using 1.7.x, allows us for doing minor releses/bugfixes that can be added to the repo, and will be automatically installed by users.

>I'm putting stress on the version number of the dependencies because it seems that's a concern in the digital preservation domain because different versions can produce different outputs. If this is going to be very hard to achieve I would also say it's ok to bump it to AM18.

The first step should be to specify the current version needed in the archivematica repo (osdeps.yml). Currently, we only list the names of packages, but if we had the version numbers, we can at least be sure that all dependencies provided by us are the expected version, but also it would be easier to detect when a package that is provided by ubuntu/centos has an update due to security reasons (AFAIK, nor ubuntu or redhat upgrade packages in their stable releases due to performance problems, only for security one)

Think about the nginx  package. Usually, it has security upgrades in each archivematica release cycle, and we can't avoid them due to the risk of security issues for our clientes. Having the current version described in osdeps/package dependencies, helps to know wich version was initially tested, and how far is the most current one afar from that.

**Comment by mamedin** [comment]
Instead of requiring the exact version we can require a minimum version:

```Requires: siegfried >= 1.7.6```

**Comment by sevein** [comment]
Ok guys thank you for explaining this to me 😅. This totally deserves its own issue.

@scollazo you mentioned in Slack that https://packages.archivematica.org/1.7.x-alpha/ is already available? Can we rename it to https://packages.archivematica.org/1.7.x/? I want to start using it here: https://github.com/artefactual/archivematica/blob/qa/1.x/src/MCPClient.Dockerfile#L26.

**Comment by sevein** [comment]
@scollazo has now published https://packages.archivematica.org/1.7.x/ - thank you!
